### ISSUE TITLE
Circular dependencies not working as with ES modules

### ISSUE THREAD

**Issue by isidrok** [systemjs/systemjs#2253]
### Demonstration

Reproducing this bug has been a bit tricky so I created a repository: https://github.com/isidrok/system-circular-dependencies-bug

The folders 'system' and 'es' contain the chunks generated by Rollup in its respective formats. As you can see, there is a circular dependency between 'core-js.js' and 'react.js' (in part created due to manually defining Rollup chunks).

Launching an http server (`npm start`) and loading both pages (http://localhost:8080/system.html and http://localhost:8080/es.html) shows that the same code works fine in ES format but doesn't work on System.

If we eliminate the circular reference caused by the file '0\commonjsHelpers.js' generated by @rollup/plugin-commonjs (maybe worth another issue?) by uncommenting the lines 26-28 of 'rollup.config.js' and building the app again (`npm run build`) then both pages work.

### Expected Behavior

Modules with circular references can be executed as its done using ES modules.

### Actual Behavior

Modules with circular references fail to execute.

**Comment by joeldenning** [comment]
Hi @isidrok, thanks for the clear description and repository demonstrating the issue.

The issue here is related to rollup, not to SystemJS. Here is the error in the browser console for the systemjs page:

<img width="542" alt="image" src="https://user-images.githubusercontent.com/5524384/94718669-4c094200-030f-11eb-8d5c-a481ec1740d1.png">

The way that rollup plugin commonjs is using the `createCommonjsModule` function is by injecting it into `c.js` and then importing it from c into b.js.

Regarding circular dependencies, the ES module spec first "links" the "hoisted exports" and then "instantiates" each module sequentially. The linking phase sets up the bindings for the exports and is what allows for the `createCommonjsModule` function to be available inside of b.js even before c.js has fully instantiated.

SystemJS supports "hoisted exports during the link phase" with the following syntax:

```js
System.register([], (_export) => {
  // link phase
  // All hoisted exports go here
  _export('foo', 'the hoisted export!');

  return {
    execute() {
      // instantiate phase
      _export('boo', 'other thing');
    }
  }
})
```

When rollup compiles to System.register format, it never does hoisted exports. You can see that this is the case by looking at this line: https://github.com/isidrok/system-circular-dependencies-bug/blob/b6e8ff9c7c4c1eddefa44d2aebc6987bd584bc03/system/c.js#L10.

To solve it, there are two options:

# Solution 1

Don't rely on hoisted exports / linking phase. To do this, you can remove the commonjs rollup plugin.

Example: https://github.com/isidrok/system-circular-dependencies-bug/pull/1

# Solution 2

Make the `createCommonjsModule` function a hoisted export.

Example: https://github.com/isidrok/system-circular-dependencies-bug/pull/2

I think that this solution is the ideal one, but it requires changes to rollup / rollup-plugin-commonjs. It might be best to open an issue up with rollup to look into it. If you do so, feel free to mention me in the rollup issue and I can help explain what needs to change.

**Comment by isidrok** [comment]
Thanks for the detailed explanation. Following my comment on the first PR and letting aside the fact that Rollup generated a circular dependency (which I did intentionally to reproduce the issue) if I understood correctly the issue is with how Rollup generates the System wrapper since hoisted exports are writen into the execute method as seen in https://github.com/rollup/rollup/blob/26d067557faba6ae6da1e130b98a12450f6328e2/src/finalisers/system.ts#L205
So they are not available before instantiation.

If that's right I'll open an issue in the Rollup repo. Thanks again!

**Comment by guybedford** [comment]
@isidrok RollupJS System output does not support circular references and I don't think it should. The argument for this is that chunk circular references always need to load together, so you should just make them the same chunk. There's no real practical reason to separate chunks across circular reference boundaries.

Perhaps we could make a warning in this case.

**Comment by isidrok** [comment]
I see, I'll explain my use case for clarification. I'm using `manualChunks` option to generate chunks for each third party library imported by a project so they can be download in parallel and being cached separately, something like:

```javascript
manualChunks(id) {
    if (id.includes('node_modules')) {
        const dirs = id.split(path.sep);
        const name = dirs[dirs.lastIndexOf('node_modules') + 1];
        return name;
    }
}
```

I was having no problems testing with ES output but when generating the same bundle for System it didn't work. A circular reference was being generated without me noticing it and that's what was causing the issue. This would not happen if letting Rollup generate its chunks.

Adding a warning at the end of the build indicating that a circular reference has been generated between some chunks would be pretty useful if the output format does not support them (maybe even an error since the code would not work). Will open an issue later.

Out of curiosity, Could it be considered that an export must be hoisted if any of its importers have a circular reference (direct or transitive) to the module exporting it, or is it more complex?

**Comment by guybedford** [comment]
Sure, I've posted a RollupJS feature request at https://github.com/rollup/rollup/issues/3801.
